openfisca_france_data/erfs_fpr/get_survey_scenario.py

```diff
@@ -1,41 +1,88 @@
+from openfisca_core.taxbenefitsystems import TaxBenefitSystem  # type: ignore
+
 from openfisca_france_data.erfs_fpr.scenario import ErfsFprSurveyScenario
-from openfisca_france_data.tests import base as base_survey
+from openfisca_france_data import france_data_tax_benefit_system
 
 
-def get_survey_scenario(year = 2012, rebuild_input_data = False, tax_benefit_system = None,
-        baseline_tax_benefit_system = None, data = None, reform_key = None):
-    assert not(
-        (tax_benefit_system is not None)
-        & (reform_key is not None)
-        )
-    if tax_benefit_system is None:
-        if reform_key is not None:
-            tax_benefit_system = base_survey.get_cached_reform(
-                reform_key = reform_key,
-                tax_benefit_system = base_survey.france_data_tax_benefit_system,
-                )
-        else:
-            tax_benefit_system = base_survey.france_data_tax_benefit_system
-
-    if baseline_tax_benefit_system is None:
-        baseline_tax_benefit_system = tax_benefit_system
+def get_survey_scenario(
+        year = 2012,
+        rebuild_input_data = False,
+        tax_benefit_system = None,
+        baseline_tax_benefit_system = None,
+        data = None,
+        reform = None
+        ):
+    '''Helper pour créer un :class:`ErfsFprSurveyScenario`.
+
+    :param year:                        L'année des données de l'enquête.
+    :param rebuild_input_data:          Si l'on doit formatter les données (raw) ou pas.
+    :param tax_benefit_system:          Le :class:`TaxBenefitSystem` déjà réformé.
+    :param baseline_tax_benefit_system: Le :class:`TaxBenefitSystem` au droit courant.
+    :param data:                        Les données de l'enquête.
+    :param reform:                      Une réforme à appliquer à *france_data_tax_benefit_system*.
+    '''
+    tax_benefit_system: TaxBenefitSystem = get_tax_benefit_system(tax_benefit_system, reform)
+    baseline_tax_benefit_system: TaxBenefitSystem = get_baseline_tax_benefit_system(baseline_tax_benefit_system)
 
     survey_scenario = ErfsFprSurveyScenario.create(
         tax_benefit_system = tax_benefit_system,
-        baseline_tax_benefit_system = tax_benefit_system,
+        baseline_tax_benefit_system = baseline_tax_benefit_system,
         year = year,
         )
+
+    # S'il n'y a pas de données, on sait où les trouver.
     if data is None:
         data = dict(
             input_data_table = dict(),
             input_data_survey_prefix = 'openfisca_erfs_fpr_data',
             )
+
+    # Les données peuvent venir en différents formats :
+    #
+    #  * Pandas' DataFrame
+    #  * CSV
+    #  * HDF5
+    #  * Etc.
+    #
+    #  .. seealso:: :func:`ErfsFprSurveyScenario.init_from_data`
     survey_scenario.init_from_data(
         data = data,
         rebuild_input_data = rebuild_input_data,
         )
     return survey_scenario
+
+
+@dispatch(TaxBenefitSystem, Reform)  # type: ignore
```

Comment:
Id do not know this decorator and it is not straightforward to understand what it does. Could you provide some ref.

Reply:
There is a lot of complexity checking whether a value is present or `None`, and that implies a lot of `if` and is error prone.

What this does is to validate data : depending on the value of the argument used, it'll _dispatch_ to a version of the function that handles that specific case.

Think of method overload in Java, for example.

What we do is to be sure to _dispatch_ to the expected behaviour by type checking the arguments.

Usually you don't need this, but I think for this repo it'll help us catch several bugs and reduce code complexity with little change.

Reply:
I can add a comment in the code for other readers as well if that helps.